File: 500.600/506.RelativeRanks.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class RelativeRanks
{
public:
    vector<string> findRelativeRanks(const vector<int> &score)
    {
        int n = score.size();
        vector<pair<int, int>> dp;
        for (int i = 0; i < n; ++i)
        {
            dp.push_back({score[i], i});
        }
        sort(dp.begin(), dp.end());
        vector<string> ans(n);
        for (int i = n - 1; i >= 0; --i)
        {
            int j = n - i;
            if (j == 1)
            {
                ans[dp[i].second] = "Gold Medal";
            }
            else if (j == 2)
            {
                ans[dp[i].second] = "Silver Medal";
            }
            else if (j == 3)
            {
                ans[dp[i].second] = "Bronze Medal";
            }
            else
            {
                ans[dp[i].second] = to_string(j);
            }
        }
        return ans;
    }
};

void print(const vector<string> &ret)
{
    cout << "[";
    for (auto &x : ret)
    {
        cout << x.c_str() << " , ";
    }
    cout << "]" << endl;
}

int main(int argc, char const *argv[])
{
    RelativeRanks cls;
    // [ "Gold Medal", "Silver Medal", "Bronze Medal", "4", "5" ]
    print(cls.findRelativeRanks(vector<int>{5, 4, 3, 2, 1}));
    // ["Gold Medal","5","Bronze Medal","Silver Medal","4"]
    print(cls.findRelativeRanks(vector<int>{10, 3, 8, 9, 4}));
    return 0;
}

File: go/500.600/506.RelativeRanks.go
package main

import (
	"fmt"
	"sort"
	"strconv"
)

type pair struct {
	a, b int
}

func findRelativeRanks(score []int) []string {
	dp := make([]*pair, len(score))
	for i, v := range score {
		dp[i] = &pair{a: v, b: i}
	}
	sort.Slice(dp, func(i, j int) bool {
		return dp[i].a > dp[j].a
	})
	ans := make([]string, len(score))
	for i, v := range dp {
		if i == 0 {
			ans[v.b] = "Gold Medal"
		} else if i == 1 {
			ans[v.b] = "Silver Medal"
		} else if i == 2 {
			ans[v.b] = "Bronze Medal"
		} else {
			ans[v.b] = strconv.Itoa(i + 1)
		}
	}
	return ans
}

func main() {
	fmt.Println(findRelativeRanks([]int{5, 4, 3, 2, 1}), `["Gold Medal","Silver Medal","Bronze Medal","4","5"]`)
	fmt.Println(findRelativeRanks([]int{10, 3, 8, 9, 4}), `["Gold Medal","5","Bronze Medal","Silver Medal","4"]`)
}
